pennylane/_qubit_device.py

```diff
@@ -29,100 +30,155 @@
 class QubitDevice(Device):
     """Abstract base class for PennyLane qubit devices.
 
+    The following abstract methods **must** be defined:
+
+    * :meth:`~.probability`: returns the probability or marginal probability from the
+      device after circuit execution. :meth:`~.marginal_prob` may be used here.
+
+    * :meth:`~.apply`: append circuit operations, compile the circuit (if applicable),
+      and perform the quantum computation.
+
+    Where relevant, devices that generate their own samples (such as hardware) should
+    overwrite the following methods:
+
+    * :meth:`~.generate_samples`: Generate samples from the device from the
+      exact or approximate probability distribution.
+
+    This device contains common utility methods for qubit-based devices. These
+    do not need to be overwritten. Utility methods include:
+
+    * :meth:`~.expval`, :meth:`~.var`, :meth:`~.sample`: return expectation values,
+      variances, and samples of observables after the circuit has been rotated
+      into the observable eigenbasis.
+
     Args:
-        wires (int): number of subsystems in the quantum state represented by the device.
-            Default 1 if not specified.
-        shots (int): Number of circuit evaluations/random samples used to estimate
-            expectation values of observables. Defaults to 1000 if not specified.
+        wires (int): number of subsystems in the quantum state represented by the device
+        shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables
+        analytic (bool): If ``True``, the device calculates probability, expectation values,
+            and variances analytically. If ``False``, a finite number of samples set by
+            the argument ``shots`` are used to estimate these quantities.
     """
-    #pylint: disable=too-many-public-methods
+
+    # pylint: disable=too-many-public-methods
+    _asarray = staticmethod(np.asarray)
+
     def __init__(self, wires=1, shots=1000, analytic=True):
         super().__init__(wires=wires, shots=shots)
+
         self.analytic = analytic
+        """bool: If ``True``, the device supports exact calculation of expectation
+        values, variance, and probabilities. If ``False``, samples are used
+        to estimate the statistical quantities above."""
+
+        self._wires_measured = set()
+        """set[int]: wires acted on by quantum operations and observables"""
 
-        self._state = None
-        self._rotated_prob = None
-        self._wires_used = None
-        self._memory = None
         self._samples = None
+        """None or array[int]: stores the samples generated by the device
+        *after* rotation to diagonalize the observables."""
 
+        self.reset()
```

Comment:
```suggestion
```

Reply:
Currently, this would be redundant (but only using ``reset`` will deteriorate the linting haha...)